presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewColumnMappingExtractor.java

```diff
@@ -0,0 +1,127 @@
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class MaterializedViewColumnMappingExtractor
+        extends MaterializedViewPlanValidator
+{
+    private final Analysis analysis;
+
+    private List<List<TableColumn>> mappedBaseColumns;
+
+    private MaterializedViewColumnMappingExtractor(Analysis analysis)
+    {
+        this.analysis = requireNonNull(analysis, "analysis is null");
+
+        this.mappedBaseColumns = new ArrayList<>();
+    }
+
+    public static List<List<TableColumn>> extractMappedBaseColumns(Analysis analysis)
+    {
+        checkState(
+                analysis.getStatement() instanceof CreateMaterializedView,
+                "Only support extracting of mapped columns from create materialized view query");
+
+        MaterializedViewColumnMappingExtractor extractor = new MaterializedViewColumnMappingExtractor(analysis);
+        Query viewQuery = ((CreateMaterializedView) analysis.getStatement()).getQuery();
+        extractor.process(viewQuery, new MaterializedViewPlanValidatorContext());
+
+        return ImmutableList.copyOf(extractor.mappedBaseColumns);
+    }
+
+    @Override
+    protected Void visitSetOperation(SetOperation node, MaterializedViewPlanValidatorContext context)
+    {
+        super.visitSetOperation(node, context);
+
+        List<RelationType> outputDescriptorList = node.getRelations().stream()
+                .map(relation -> analysis.getOutputDescriptor(relation).withOnlyVisibleFields())
+                .collect(toImmutableList());
+
+        int numRelations = outputDescriptorList.size();
+        int numFields = outputDescriptorList.get(0).getVisibleFieldCount();
+        for (int fieldIndex = 0; fieldIndex < numFields; fieldIndex++) {
+            for (int firstRelationIndex = 0; firstRelationIndex < numRelations; firstRelationIndex++) {
+                Optional<TableColumn> firstBaseColumn = tryGetOriginalTableColumn(outputDescriptorList.get(firstRelationIndex).getFieldByIndex(fieldIndex));
+                for (int secondRelationIndex = firstRelationIndex + 1; secondRelationIndex < numRelations; secondRelationIndex++) {
+                    Optional<TableColumn> secondBaseColumn = tryGetOriginalTableColumn(outputDescriptorList.get(secondRelationIndex).getFieldByIndex(fieldIndex));
+                    if (firstBaseColumn.isPresent() && secondBaseColumn.isPresent() && !firstBaseColumn.get().equals(secondBaseColumn.get())) {
+                        mappedBaseColumns.add(ImmutableList.of(firstBaseColumn.get(), secondBaseColumn.get()));
+                    }
+                }
+            }
+        }
+
+        return null;
+    }
+
+    @Override
+    protected Void visitComparisonExpression(ComparisonExpression node, MaterializedViewPlanValidatorContext context)
```

Comment:
I feel like this is basically a customized equivalence class computation which only works for join conjuncts. I am wondering if there are any generic equivalence class generation utils available in presto which can be reused here? @highker 
Do we also want to handle the cases where the ComparisonExpression is in the where clause instead of the join conjuncts?

Reply:
Regarding
```
Do we also want to handle the cases where the ComparisonExpression is in the where clause instead of the join conjuncts?
```
Better to have but we can get by at the moment by only taking care of the join criteria. If we fail to capture a single partition mapping by not checking on the where clause, the query shape validation will fail during create mv. On the other hand, as long as we are able to capture at least one partition mapping by checking only the join criteria, it is sufficient to get the select & refresh logic correct, even though not fine grained and perfect.